File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiRibbonGroup.cpp
#include "StdAfx.h"
#include "UiRibbonGroup.h"

#define ThisMethod($x) &UiRibbonGroup::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiRibbonGroup::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiRibbonGroup> c_obj;
	}

	void UiRibbonGroup::DefineControl()
	{
		AutoAddMethod( SetText );
		AutoAddMethod( GetText, WrapObjectGeneric );
		AutoAddMethod( SetDetailButton );
		AutoAddMethod( GetDetailButton, WrapObjectGeneric );
		AutoAddMethod( SetVisual );
		AutoAddMethod( GetVisual, WrapObjectGeneric );
		AutoAddMethod( ControlAdd );
		AutoAddMethod( ControlInsert );
		AutoAddMethod( ControlGetByIndex, WrapObjectGeneric );
		AutoAddMethod( ControlGetById, WrapObjectGeneric );
		AutoAddMethod( ControlRemoveByIndex );
		AutoAddMethod( ControlRemoveById );
		AutoAddMethod( ControlRemove );
		AutoAddMethod( ControlRemoveAll );
		AutoAddMethod( OnDetailClick, WrapObjectGeneric );
	}

	U1 UiRibbonGroup::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}

	void UiRibbonGroup::__OnDetailClick( Ui::IRibbonGroup & sender )
	{
		m_OnClick( this );
	}

	UiRibbonGroup* UiRibbonGroup::OnDetailClick( Callback_t&& fn ) { m_OnClick = SetEventCallback<Ui::IRibbonGroup::OnDetailClick>( std::move( fn ), MakeThisFunc( __OnDetailClick ) ); return this; }

}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Ribbon/UiRibbonGroup.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { WindowLike } from "../Control/UiWindow";
import { IVisual } from "../UiVisual";

export interface IRibbonGroup extends IControl {
	new (window: WindowLike, key?: string | number): IRibbonGroup;

	SetText(s: string): RibbonGroup;
	GetText(): string;

	SetDetailButton(bDetailButton: boolean): RibbonGroup;
	GetDetailButton(): boolean;

	SetVisual(pv: IVisual): IVisual;
	GetVisual(): IVisual;

	ControlAdd(pChild: IControl): boolean;
	ControlInsert(nInsertBefore: number, pChild: IControl): boolean;
	ControlGetByIndex(nIndex: number): IControl;
	ControlGetById(nId: number): IControl;
	ControlRemoveByIndex(nIndex: number): IControl;
	ControlRemoveById(nId: number): IControl;
	ControlRemove(pChild: IControl): IControl;
	ControlRemoveAll(): RibbonGroup;

	OnDetailClick(fn: (sender: RibbonGroup) => void): RibbonGroup;
}

export class RibbonGroup extends (AveLib.UiRibbonGroup as IRibbonGroup) {
	private m_Visual: IVisual;
	private m_Control: Set<IControl> = new Set();

	SetVisual(pv: IVisual) {
		this.m_Visual = pv;
		return super.SetVisual(pv);
	}

	ControlAdd(pChild: IControl) {
		this.m_Control.add(pChild);
		return super.ControlAdd(pChild);
	}

	ControlInsert(nInsertBefore: number, pChild: IControl) {
		this.m_Control.add(pChild);
		return super.ControlInsert(nInsertBefore, pChild);
	}

	ControlRemoveByIndex(nIndex: number) {
		const t = super.ControlRemoveByIndex(nIndex);
		this.m_Control.delete(t);
		return t;
	}

	ControlRemoveById(nId: number) {
		const t = super.ControlRemoveById(nId);
		this.m_Control.delete(t);
		return t;
	}

	ControlRemove(pChild: IControl) {
		this.m_Control.delete(pChild);
		return super.ControlRemove(pChild);
	}

	ControlRemoveAll() {
		this.m_Control.clear();
		return super.ControlRemoveAll();
	}
}
